resources/persistence.rb

```diff
@@ -0,0 +1,151 @@
+property :alt_names, Array
+property :key,       String, required: true
+property :crt,       String
+property :chain,     String
+property :fullchain, String
+
+property :master, [TrueClass, FalseClass], default: false
+
+property :data_bag_name, String, required: true
+property :encrypt, [TrueClass, FalseClass], default: true
+property :secret, String
+
+property :owner,     String, default: 'root'
+property :group,     String, default: 'root'
+
+action :save do
+  unless master
+    Chef::Log.warn "master property not set, will not save #{cn} certificates"
+    return
+  end
+
+  data = {
+    'id'         => cn,
+    'alt_names'  => alt_names,
+    'created_by' => node['fqdn'],
+    'created_at' => Time.now
+  }
+
+  # 'key', 'cert', 'chain' are also used in the data bag format used by
+  # https://github.com/atomic-penguin/cookbook-certificate/blob/master/providers/manage.rb
+  data['key']       = ::File.read(new_resource.key)       if new_resource.key
+  data['cert']      = ::File.read(new_resource.crt)       if new_resource.crt
+  data['chain']     = ::File.read(new_resource.chain)     if new_resource.chain
+
+  data['fullchain'] = ::File.read(new_resource.fullchain) if new_resource.fullchain
+
+  chef_data_bag_item "#{data_bag_name}/#{cn}" do
+    raw_data data
+    if new_resource.encrypt && (new_resource.secret || default_data_bag_secret)
+      encrypt true
+      encryption_version 2
+      secret new_resource.secret || default_data_bag_secret
+    end
+  end
+end
+
+# Matrix:
+#
+# +------------------------+-----------------+--------------------------------------+
+# |  file                  |  data bag item  |  action                              |
+# | ---------------------- | --------------- | ------------------------------------ |
+# |  does not exist        |  exists         |  create from data bag item           |
+# |  does not exist        |  does not exist |  nothing (-> acme_selfsigned)        |
+# |  exists (self-signed)  |  exists         |  create from data bag item           |
+# |  exists (self-signed)  |  does not exist |  nothing (~> renew acme_certificate) |
+# |  exists (valid)        |  is newer       |  create from data bag item           |
+# |  exists (valid)        |  is older       |  nothing (~> renew acme_certificate) |
+# |  exists (expired)      |  is newer       |  create from data bag item           |
+# |  exists (expired)      |  does not exist |  nothing (~> renew acme_certificate  |
+# +------------------------+-----------------+--------------------------------------+
+#
+action :load do
+  begin
+    existing_cert = ::OpenSSL::X509::Certificate.new(::File.read(crt || fullchain))
+  rescue Errno::ENOENT => e
+    Chef::Log.warn("certificate file #{crt || fullchain} does not exist yet: #{e}")
+  rescue OpenSSL::X509::CertificateError => e
+    Chef::Log.error("certificate file #{crt || fullchain} exists but is broken: #{e}")
+  end
+
+  item = load_data_bag_item(data_bag_name, 'id:' + cn, secret)
```

Comment:
I think you don’t need this custom method, and can instead use the built-in `data_bag_item()` method, since you’re in a scope that has it.

Reply:
My solution was supposed to support three cases:

- no encrypted data bag -> empty secret
- custom data bag secret -> manual set secret
- default data bag secret ->  manual set secret

using `data_bag_item` without a secret automatically falls back to the secret defined in `Chef::Config[:encrypted_data_bag_secret]` and iirc providing `nil` as secret throws an exception

Reply:
@rmoriz ah! I forgot about the `nil` throwing an exception. In that case, this is great.